io/src/main/java/io/smallrye/common/io/PathUtils.java

```diff
@@ -0,0 +1,2139 @@
+     *         path
+     * @see <a href=
+     *      "https://github.com/google/guava/blob/master/guava/src/com/google/common/io/MoreFiles.java">com.google.common.io.MoreFiles</a>
+     */
+    public static Path deleteRecursively(final Path path) throws UncheckedIOException {
+        return doDeleteRecursively(checkExistsNoFollowLinks(path));
+    }
+
+    static Path doDeleteRecursively(final Path path) throws UncheckedIOException {
+        Path parentPath = getParentPath(path);
+        if (parentPath == null) {
+            throw new UncheckedIOException(new FileSystemException(path.toString(), null, "can't delete recursively"));
+        }
+
+        Collection<IOException> exceptions = null; // created lazily if needed
+        try {
+            boolean sdsSupported = false;
+            try (DirectoryStream<Path> parent = Files.newDirectoryStream(parentPath)) {
+                if (parent instanceof SecureDirectoryStream) {
+                    sdsSupported = true;
+                    exceptions = deleteRecursivelySecure((SecureDirectoryStream<Path>) parent,
+                            /*
+                             * requireNonNull is safe because paths have file names when they have parents,
+                             * and we checked for a parent at the beginning of the method.
+                             */
+                            checkNotNullParam("fileName", path.getFileName()), exceptions);
+                }
+            }
+
+            if (!sdsSupported) {
+                exceptions = deleteRecursivelyInsecure(path, exceptions);
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+
+        if (exceptions != null) {
+            FileSystemException deleteFailed = new FileSystemException(path.toString(), null,
+                    "failed to delete one or more files; see suppressed exceptions for details");
+            for (IOException e : exceptions) {
+                deleteFailed.addSuppressed(e);
+            }
+            throw new UncheckedIOException(deleteFailed);
+        }
+        return path;
+    }
+
+    /**
+     * Recursive deletion of directories content, including subfolders.
+     *
+     * Adopted from Google Guava.
+     *
+     * @param path to delete, must be not null, a path, and exists.
+     * @return the path to delete
+     * @throws UncheckedIOException around original IOExceptions and if not exists
+     *         path
+     * @see <a href=
+     *      "https://github.com/google/guava/blob/master/guava/src/com/google/common/io/MoreFiles.java">com.google.common.io.MoreFiles</a>
+     */
+    public static Path deleteDirectoryContents(Path path) throws UncheckedIOException {
+        return doDeleteDirectoryContents(checkIsDirectoryNoFollowLinks(path));
+    }
+
+    static Path doDeleteDirectoryContents(Path path) throws UncheckedIOException {
+        Collection<IOException> exceptions = null; // created lazily if needed
```

Comment:
This is minor but you could have this variable be of type `FileSystemException` and pass that in to the nested methods rather than adding the suppressed exceptions to a collection only to iterate them all right away.

Reply:
The current approach allows lazy init. If I change that, the top-most exception would have a stacktrace top of a utility method or I can't lazy init.